pbsmetrics/prometheus/prometheus.go

```diff
@@ -0,0 +1,242 @@
+		"Seconds to resolve each PBS request.",
+		standardLabelNames, timerBuckets,
+	)
+	metrics.Registry.MustRegister(metrics.reqTimer)
+	metrics.adaptRequests = newCounter(cfg, "adapter_requests_total",
+		"Number of requests sent out to each bidder.",
+		adapterLabelNames,
+	)
+	metrics.Registry.MustRegister(metrics.adaptRequests)
+	metrics.adaptTimer = newHistogram(cfg, "adapter_time_seconds",
+		"Seconds to resolve each request to a bidder.",
+		adapterLabelNames, timerBuckets,
+	)
+	metrics.Registry.MustRegister(metrics.adaptTimer)
+	metrics.adaptBids = newCounter(cfg, "adapter_bids_recieved_total",
+		"Number of bids recieved from each bidder.",
+		bidLabelNames,
+	)
+	metrics.Registry.MustRegister(metrics.adaptBids)
+	metrics.adaptPrices = newHistogram(cfg, "adapter_prices",
+		"Value of the highest bids from each bidder.",
+		adapterLabelNames, prometheus.LinearBuckets(0.1, 0.1, 200),
+	)
+	metrics.Registry.MustRegister(metrics.adaptPrices)
+	metrics.cookieSync = newCookieSync(cfg)
+	metrics.Registry.MustRegister(metrics.cookieSync)
+	metrics.userID = newCounter(cfg, "usersync_total",
+		"Number of user ID syncs performed",
+		[]string{"action", "bidder"},
+	)
+	metrics.Registry.MustRegister(metrics.userID)
+
+	return &metrics
+}
+
+func newConnCounter(cfg config.PrometheusMetrics) prometheus.Gauge {
+	opts := prometheus.GaugeOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      "active_connections",
+		Help:      "Current number of active (open) connections.",
+	}
+	return prometheus.NewGauge(opts)
+}
+
+func newCookieSync(cfg config.PrometheusMetrics) prometheus.Counter {
+	opts := prometheus.CounterOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      "cookie_sync_requests_total",
+		Help:      "Number of cookie sync requests recieved.",
+	}
+	return prometheus.NewCounter(opts)
+}
+
+func newCounter(cfg config.PrometheusMetrics, name string, help string, labels []string) *prometheus.CounterVec {
+	opts := prometheus.CounterOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+	}
+	return prometheus.NewCounterVec(opts, labels)
+}
+
+func newHistogram(cfg config.PrometheusMetrics, name string, help string, labels []string, buckets []float64) *prometheus.HistogramVec {
+	opts := prometheus.HistogramOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+		Buckets:   buckets,
+	}
+	return prometheus.NewHistogramVec(opts, labels)
+}
+
+func (me *Metrics) RecordConnectionAccept(success bool) {
+	if success {
+		me.connCounter.Inc()
+	} else {
+		me.connError.WithLabelValues("accept_error").Inc()
+	}
+
+}
+
+func (me *Metrics) RecordConnectionClose(success bool) {
+	if success {
+		me.connCounter.Dec()
+	} else {
+		me.connError.WithLabelValues("close_error").Inc()
+	}
+}
+
+func (me *Metrics) RecordRequest(labels pbsmetrics.Labels) {
+	me.requests.With(resolveLabels(labels)).Inc()
```

Comment:
so... one major warning here. The implementation on this `With()` function ultimately handles a lock:

```
// getOrCreateMetricWithLabelValues retrieves the metric by hash and label value
// or creates it and returns the new one.
//
// This function holds the mutex.
func (m *MetricVec) getOrCreateMetricWithLabels(hash uint64, labels Labels) Metric {
	m.mtx.RLock()
	metric, ok := m.getMetricWithLabels(hash, labels)
	m.mtx.RUnlock()
```

I don't think this should be handled in this PR... but if we see weird issues (especially on startup, when all these metric/label combinations are being created), we might have to be smarter about this.

I've been warned that global locks will wreck performance in these high-traffic systems... so it might be worth opening an issue about, and logging with a `// TODO #XYZ` comment so that we don't forget.

In the future (assuming you remove `publisher_id` from the labels), we could call this function with all the label combinations on app startup so that we're only grabbing read-locks while serving real traffic.